### PR TITLE
LID benchmarking tool

### PR DESCRIPTION
**Pull request by dirkmc**
A simple tool to measure performance of the various solutions we're looking at for LID:
- [x] Yugabyte DB
  - [x] SQL tables
  - [x] Cassandra
- [x] Postgres
- [ ] FoundationDB
- [x] Cassandra

Note: At the moment the implementations are for a simple one-machine load test.
We will need to do some work to be able to run these load tests at scale.

```
$ go run ./bench -vv postgres --piece-count=10 --bs-fetch-count=1000 -bs-fetch-parallelism=10 --piece-add-parallelism=4
2023-03-14T15:46:21.287+0100	INFO	bench	bench/run.go:48	Running benchmark for Postgres DB
2023-03-14T15:46:21.287+0100	INFO	bench	bench/run.go:49	Initializing...
2023-03-14T15:46:21.464+0100	INFO	bench	bench/run.go:53	Initialized
2023-03-14T15:46:21.464+0100	INFO	bench	bench/run.go:88	Adding 10 x 1.0 MB pieces...
2023-03-14T15:46:22.087+0100	INFO	bench	bench/run.go:174	Added 10 pieces in 622.965466ms
2023-03-14T15:46:22.087+0100	DEBUG	bench	bench/run.go:175	total CPU time 2.251036128s
2023-03-14T15:46:22.087+0100	DEBUG	bench	bench/run.go:176	  created random files in 82.062981ms
2023-03-14T15:46:22.087+0100	DEBUG	bench	bench/run.go:177	  created CARs in 911.662939ms
2023-03-14T15:46:22.087+0100	DEBUG	bench	bench/run.go:178	  read records in 281.433904ms
2023-03-14T15:46:22.087+0100	DEBUG	bench	bench/run.go:179	  added records in 975.876304ms
2023-03-14T15:46:22.087+0100	INFO	bench	bench/run.go:185	Bitswap simulation: fetching 1000 random blocks with parallelism 10...
2023-03-14T15:46:22.124+0100	INFO	bench	bench/run.go:230	Bitswap simulation completed in 37.085109ms
2023-03-14T15:46:22.125+0100	DEBUG	bench	bench/run.go:231	total CPU time: 204.124342ms
2023-03-14T15:46:22.125+0100	DEBUG	bench	bench/run.go:232	  multihash lookup total time: 198.245519ms
2023-03-14T15:46:22.125+0100	DEBUG	bench	bench/run.go:233	  get offset / size total time: 5.878823ms
2023-03-14T15:46:22.125+0100	INFO	bench	bench/run.go:239	Graphsync simulation: fetching 10 random blocks with parallelism 3...
2023-03-14T15:46:22.144+0100	INFO	bench	bench/run.go:284	Graphsync simulation completed in 19.566744ms
2023-03-14T15:46:22.144+0100	DEBUG	bench	bench/run.go:285	total CPU time: 21.976998ms
2023-03-14T15:46:22.144+0100	DEBUG	bench	bench/run.go:286	  multihash lookup total time: 13.491169ms
2023-03-14T15:46:22.144+0100	DEBUG	bench	bench/run.go:287	  get index total time: 8.485829ms
```

### CODE REVIEW

File: extern/boostd-data/bench/run.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+					totalCreateRecs += time.Since(createRecsStart)
+
+					// Add the records to the db
+					addRecsStart := time.Now()
+					pcid := testutil.GenerateCid()
+					err := db.AddIndexRecords(ctx, pcid, recs)
+					if err != nil {
+						return err
+					}
+					totalAddRecs += time.Since(addRecsStart)
```
Comment (magik6k) [inline extern/boostd-data/bench/run.go:147]: Not thread safe, should either use atomic.Add or locks

File: extern/boostd-data/bench/run.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+						recs = append(recs, model.Record{
+							Cid: c,
+							OffsetSize: model.OffsetSize{
+								Offset: uint64(rand.Intn(sectorSize)),
+								Size:   uint64(rand.Intn(sectorSize)),
+							},
+						})
+					}
+					totalCreateRecs += time.Since(createRecsStart)
```
Comment (magik6k) [inline extern/boostd-data/bench/run.go:138]: Also not thread safe

File: extern/boostd-data/bench/run.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+	err := eg.Wait()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Added %d pieces in %s", pieceCount, time.Since(addStart))
+	log.Debugf("total CPU time %s", totalCreateRecs+totalAddRecs)
+	log.Debugf("  created records in %s", totalCreateRecs)
+	log.Debugf("  added records in %s", totalAddRecs)
```
Comment (magik6k) [inline extern/boostd-data/bench/run.go:161]: Would use `log.*w` for easier to parse structured logs

File: extern/boostd-data/bench/run.go (hunk 1 above)
Comment (magik6k) [inline extern/boostd-data/bench/run.go:147]: Would also log time+record count here so it's possible to see perf over time, and see how batch size / parallelism / etc. impact individual batch perf

File: extern/boostd-data/bench/postgres.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		if i > 0 {
+			vals = vals + ","
+		}
+		vals = vals + fmt.Sprintf("($%d,$%d)", (i*2)+1, (i*2)+2)
+		args = append(args, rec.Cid.Hash(), pieceCid.Bytes())
+	}
+	_, err = tx.ExecContext(ctx, `INSERT INTO PayloadToPieces (PayloadMultihash, PieceCids) VALUES `+vals, args...)
+	if err != nil {
+		return fmt.Errorf("executing insert: %w", err)
+	}
```
Comment (magik6k) [inline extern/boostd-data/bench/postgres.go:181]: This is not the fastest way to go about doing large inserts to postgres, especially if you have some latency.

In my experience the fastest way, by far, is `copy .. from stdin` into a temp table, then insert-select from that table ignoring conflicts.

See e.g. https://github.com/filecoin-project/lotus/blob/v1.4.2/cmd/lotus-chainwatch/processor/messages.go#L121-L159

File: extern/boostd-data/bench/run.go (hunk 1 above)
Comment (dirkmc) [reply extern/boostd-data/bench/run.go:153]: We're planning to test with 100s of millions of records so we'll probably need to do something in an aggregate fashion

File: extern/boostd-data/bench/cassandra.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+			err := c.session.ExecuteBatch(batch)
+			if err != nil {
+				return fmt.Errorf("adding index records for piece %s: %w", pieceCid, err)
+			}
+			batchIdx = 0
+			continue
+		}
+
+		batch.Entries = append(batch.Entries, entry)
+		batchIdx++
```
Comment (magik6k) [inline extern/boostd-data/bench/cassandra.go:174]: This code is moderately confusing, to me this reads like the last batch entry will never be it a batch that is executed.

The append should probably be before this if.

File: extern/boostd-data/bench/bench
Comment (magik6k) [inline extern/boostd-data/bench/bench:1]: probably don't want that in git history

**Review by nonsense — APPROVED** [review]

**Pull request merged**